code/datums/components/trader/trader.dm

```diff
@@ -0,0 +1,413 @@
+		TRADER_RADIAL_LORE = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_lore"),
+		TRADER_RADIAL_DISCUSS_BUY = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_buying"),
+		TRADER_RADIAL_DISCUSS_SELL = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_selling"),
+		TRADER_RADIAL_YES = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_yes"),
+		TRADER_RADIAL_NO = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_no"),
+		TRADER_RADIAL_OUT_OF_STOCK = image(icon = 'icons/hud/radial.dmi', icon_state = "radial_center"),
+	)
+
+	restock_products()
+	renew_item_demands()
+
+/datum/component/trader/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ATOM_ATTACK_HAND, PROC_REF(on_attack_hand))
+
+/datum/component/trader/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_ATOM_ATTACK_HAND)
+
+///If our trader is alive, and the customer left clicks them with an empty hand without combat mode
+/datum/component/trader/proc/on_attack_hand(atom/source, mob/living/carbon/customer)
+	SIGNAL_HANDLER
+	var/mob/living/living_trader = parent
+	if(!can_trade(customer) || customer.combat_mode)
+		return
+	var/list/npc_options = list()
+	if(products.len)
+		npc_options["Buy"] = radial_icons_cache[TRADER_RADIAL_BUY]
+	if(wanted_items.len)
+		npc_options["Sell"] = radial_icons_cache[TRADER_RADIAL_SELL]
+	if(length(trader_data.say_phrases))
+		npc_options["Talk"] = radial_icons_cache[TRADER_RADIAL_TALK]
+	if(!npc_options.len)
+		return
+
+	living_trader.face_atom(customer)
+
+	INVOKE_ASYNC(src, PROC_REF(open_npc_options), customer, npc_options)
+
+	return COMPONENT_CANCEL_ATTACK_CHAIN
+
+/**
+ * Generates a radial of the initial radials of the NPC
+ * Called via asynch, due to the sleep caused by show_radial_menu
+ * Arguments:
+ * * customer - (Mob REF) The mob trying to buy something
+ */
+/datum/component/trader/proc/open_npc_options(mob/living/carbon/customer, list/npc_options)
+	if(!can_trade(customer))
+		return
+	var/npc_result = show_radial_menu(customer, parent, npc_options, custom_check = CALLBACK(src, PROC_REF(check_menu), customer), require_near = TRUE, tooltips = TRUE)
+	switch(npc_result)
+		if("Buy")
+			buy_item(customer)
+		if("Sell")
+			try_sell(customer)
+		if("Talk")
+			discuss(customer)
+
+/**
+ * Checks if the customer is ok to use the radial
+ *
+ * Checks if the customer is not a mob or is incapacitated or not adjacent to the source of the radial, in those cases returns FALSE, otherwise returns TRUE
+ * Arguments:
+ * * customer - (Mob REF) The mob checking the menu
+ */
+/datum/component/trader/proc/check_menu(mob/customer)
+	if(!istype(customer))
+		return FALSE
+	if(IS_DEAD_OR_INCAP(customer) || !customer.Adjacent(parent))
+		return FALSE
+	return TRUE
+
+/**
+ * Generates a radial of the items the NPC sells and lets the user try to buy one
+ * Arguments:
+ * * customer - (Mob REF) The mob trying to buy something
+ */
+/datum/component/trader/proc/buy_item(mob/customer)
```

Comment:
add some newlines in this proc it's very hard to read

Reply:
I have separated a the if blocks to make them easier on the eye, hopefully I have not gone overboard.